### PR TITLE
feat: add string transform functions

### PR DESCRIPTION
**Pull request by richtia**
PR to add definitions for string transform functions.

### CODE REVIEW

File: extensions/functions_string.yaml
Comment (richtia) [inline extensions/functions_string.yaml:259]: I'm unsure if this is the proper format for variadic.  It should only apply to the "string_arguments"

File: extensions/functions_string.yaml
Comment (richtia) [inline extensions/functions_string.yaml:343]: I was mainly looking here when writing this:
https://forums.mysql.com/read.php?103,187048,188748#msg-188748
https://dev.mysql.com/doc/refman/5.7/en/charset-unicode-sets.html

```
MySQL implements the xxx_unicode_ci collations according to the Unicode Collation Algorithm (UCA) described at http://www.unicode.org/reports/tr10/
```

Follow up to @jacques-n comment here:
https://github.com/substrait-io/substrait/pull/245#discussion_r921711088

Not sure if these are the kinds of rules we're looking for, so open to more discussion.

File: extensions/functions_string.yaml
Comment (jvanstraten) [reply extensions/functions_string.yaml:259]: The variadic key is pary of the implemenation structure, not the argument structure. It always applies to the last argument, so only it can be made variadic.

Comment (richtia) [reply extensions/functions_string.yaml]: updated thanks!

**Review by jvanstraten — CHANGES_REQUESTED** [review]

File: extensions/functions_string.yaml
Comment (richtia) [reply extensions/functions_string.yaml:429]: Yeah, it seems to be bits lol.  https://www.w3resource.com/PostgreSQL/bit-length-function.php

File: extensions/functions_string.yaml
Comment (richtia) [reply extensions/functions_string.yaml:343]: I added the option, let me know if you think it's okay.

File: extensions/functions_string.yaml
Comment (richtia) [reply extensions/functions_string.yaml]: good catch.  modified

File: extensions/functions_string.yaml
Comment (jvanstraten) [reply extensions/functions_string.yaml:343]: I can't really fault you for this because in all likelihood nothing says anything about this, but I'd strongly prefer we use underscores for the option names. Dashes are an absolute PITA to parse in any DSL because they're ambiguous with subtractions. If there's precedent for dashes you can just leave it as is though, I guess...

File: extensions/functions_string.yaml
Comment (jvanstraten) [reply extensions/functions_string.yaml]: The docs there don't list any reason why it's in bits rather than bytes... IMO it should just not exist in this form and should just be `byte_length`. When converting a `bit_length(x)` function in SQL to Substrait it can always be represented as `multiply(byte_length(x), 8)`, so `byte_length` is more primitive. I could see the reason they used bit granularity being that with different storage compression methods you might end up with bit-level stuff, but Substrait doesn't define any of that stuff for `string`s; all it says is that it's UTF-8-encoded, so at best you can have a function that specifies the number of bytes needed for said UTF-8 encoding.

OTOH, there's precedent for following SQL spec nonsense (like `sum([])` returning null). Is this specifically a Postgres function or is it part of the SQL spec?

Comment (richtia) [reply extensions/functions_string.yaml]: Postgres and MySQL have bit_length, they both also have one called 'octet_length', which returns bytes: 
https://www.w3resource.com/PostgreSQL/octet_length-function.php

I can add a separate byte_length function, so we have both.

Comment (jvanstraten) [reply extensions/functions_string.yaml]: > The above function is a synonym for LENGTH().

Are their strings not UTF-8? I don't even know anymore.

I think I'm just going to abstain from this discussion because I increasingly feel like I don't have the background to make informed decisions about this. @jacques-n do you hold strong opinions about these functions?

Comment (richtia) [reply extensions/functions_string.yaml]: Yeah, it's confusing because i've seen length defined 2 different ways.  One returns the bytes, the other returns the number of characters (which is why I just updated the length function in this PR to be char_length).

File: extensions/functions_string.yaml
Comment (richtia) [reply extensions/functions_string.yaml:343]: Thanks for the catch...I think all the other options actually do use underscores.  I just changed it.

**Review by jvanstraten — DISMISSED** [review]
LGTM if no one has any further input on the open `bit_length` discussion.

File: extensions/functions_string.yaml
Comment (richtia) [reply extensions/functions_string.yaml]: @jacques-n SQL Server and Oracle have both bit_length and octet_length (for bytes).  I've added octet_length to this PR as well.

https://docs.oracle.com/cd/B19188_01/doc/B15917/sqfunc.htm#i1005968
https://docs.microsoft.com/en-us/sql/odbc/reference/appendixes/string-functions?view=sql-server-ver16

**Review by jvanstraten — APPROVED** [review]
LGTM

**Comment by richtia** [comment]
> LGTM

Just made the change so that concat is also variadic

**Review by jvanstraten — APPROVED** [review]

**Comment by ianmcook** [comment]
I think `concat` and `concat_ws` should accept a minimum of _one_ argument (not two). PostgreSQL, MySQL, and other systems allow that.

**Comment by richtia** [comment]
> I think `concat` and `concat_ws` should accept a minimum of _one_ string argument (not two). PostgreSQL, MySQL, and other systems allow that.

Updated.  Thanks!

**Comment by jvanstraten** [comment]
FWIW I accepted it as 2 minimum because 1 reduces to no operation, but I don't feel strongly either way. Arguably it further generalizes to returning the empty string for 0 arguments.

**Review by jvanstraten — DISMISSED** [review]

**Comment by richtia** [comment]
Rebased and added a bunch missing function argument names

**Review by jvanstraten — APPROVED** [review]

**Pull request merged**
